### PR TITLE
Support simpler custom interceptors for upgraded properties

### PR DESCRIPTION
**Pull request by asodja**
This implements simpler interceptors for bytecode upgrades. You can now add an interceptor like:

```
@ReplacesEagerProperty(adapter = ConfigurationsPropertyAdapter.class)
public abstract SetProperty<String> getConfigurations();

static class ConfigurationsPropertyAdapter {
    @BytecodeUpgrade
    static Set<Configuration> getConfigurations(AbstractDependencyReportTask thiz) {
         ...
    }

    @BytecodeUpgrade
    static void setConfigurations(AbstractDependencyReportTask thiz, Set<Configuration> configurations) {
        ...
    }
}
```
We also validate that the class is package private. And then we generate a `public` `$$BridgeFor$$` class with `public` methods, so they can be accessible from any call-site. These methods are deprecated.

Additionally we also automatically accept binary compatibility changes for upgraded methods. 

I used `adapter` instead of `adaptor`, since we use `adapter` version of a word in some other place for bytecode upgrade, but we can also rename that if preferred. 

Fixes https://github.com/gradle/gradle/issues/29363

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java [hunk 1]
```diff
@@ -146,6 +157,13 @@ public Collection<Result> readRequest(ExecutableElement method) {
 
     @SuppressWarnings("unchecked")
     private List<AccessorSpec> readAccessorSpecsFromReplacesEagerProperty(ExecutableElement method, AnnotationMirror annotationMirror) {
+        Element element = AnnotationUtils.findAnnotationValueWithDefaults(elements, annotationMirror, "adapter")
+            .map(v -> types.asElement((TypeMirror) v.getValue()))
+            .orElseThrow(() -> new IllegalArgumentException("Missing adapter value"));
+        if (!element.getSimpleName().toString().equals(DefaultValue.class.getSimpleName())) {
+            return readAccessorSpecsFromAdapter(element, method.getEnclosingElement(), annotationMirror);
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:164]: 💭 Here we read an adapter from annotation and then in `readAccessorSpecsFromAdapter` read methods from adapter

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java [hunk 2]
```diff
@@ -162,6 +180,90 @@ private List<AccessorSpec> readAccessorSpecsFromReplacesEagerProperty(Executable
...
+            .filter(method -> method.getAnnotation(BytecodeUpgrade.class) != null)
+            .collect(Collectors.toList());
+        validateBridgedMethods(adapter, upgradedElement, bridgedMethods);
+
+        return bridgedMethods.stream()
+            .map(method -> bridgedMethodToAccessorSpec(method, annotationMirror))
+            .collect(Collectors.toList());
+    }
+
+    private static void validateBridgedMethods(Element adapter, Element upgradedElement, List<ExecutableElement> methods) {
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:194]: 💭 Here we do some validation that Adapter is package private etc.

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java [hunk 3]
```diff
@@ -78,26 +85,67 @@ private static MethodSpec mapToMethodSpec(CallInterceptionRequest request, Consu
...
+
             onProcessedRequest.accept(request);
             return spec;
         } catch (Exception e) {
             onFailure.accept(new HasFailures.FailureInfo(request, e.getMessage()));
             throw e;
         }
+    }
+
+    private static MethodSpec mapToBridgedMethod(String methodName, PropertyUpgradeRequestExtra implementationExtra, CallableInfo callable) {
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java:116]: 💭 Here is a part where we generate `$$BridgeFor$$` methods that are public and delegate to Adapter methods

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java [hunk 4]
```diff
@@ -78,26 +85,67 @@ private static MethodSpec mapToMethodSpec(CallInterceptionRequest request, Consu
...
+        List<TypeName> exceptions = bridgedMethod.getThrownTypes().stream()
+            .map(TypeName::get)
+            .collect(Collectors.toList());
+        String passedParameters = parameters.stream()
+            .map(parameterSpec -> parameterSpec.name)
+            .collect(Collectors.joining(", "));
+
+        CodeBlock.Builder bodyBuilder = CodeBlock.builder();
+        if (implementationExtra.getDeprecationSpec().isEnabled()) {
+            bodyBuilder.addStatement(getDeprecationCodeBlock(implementationExtra, callable));
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java:133]: 💭 For `$$BridgeFor$$` class methods we also generate deprecation

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/modelreader/impl/TypeUtils.java [hunk 5]
```diff
@@ -36,4 +44,29 @@ public static String extractMethodDescriptor(ExecutableElement methodElement) {
             methodElement.getParameters().stream().map(it -> extractType(it.asType())).toArray(Type[]::new)
         );
     }
+
+    @Nonnull
+    public static List<ExecutableElement> getExecutableElementsFromElements(Stream<? extends Element> elements) {
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/modelreader/impl/TypeUtils.java:49]: 💭 These two methods here are just moved from `AbstractInstrumentationProcessor`, so we can reuse them when we read methods on `Adapter`

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        when:
+        Compilation compilation = compile(givenSource)
+
+        then:
+        def generatedClass = source """
+             package org.gradle.test;
+             import org.gradle.api.Generated;
+
+             @Generated
+             public final class \$\$BridgeFor\$\$Task\$\$TaskAdapter {
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy:71]: 💭 This test shows what kind of `$$BridgeFor$$` class gets generated

**Comment by asodja** [comment]
@bot-gradle test this

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy [hunk 7]
```diff
@@ -50,6 +51,7 @@ class BinaryCompatibilityHelper {
         japicmpTask.tap {
             addExcludeFilter(AnonymousClassesFilter)
             addExcludeFilter(KotlinInternalFilter)
+            addExcludeFilter(BridgeForClassFilter)
```
Comment (asodja) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy:54]: 💭 We filter $$BridgeFor$$ classes for binary compatibilty check

**Comment by asodja** [comment]
@bot-gradle test this

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy [hunk 8]
```diff
@@ -50,6 +51,7 @@ class BinaryCompatibilityHelper {
         japicmpTask.tap {
             addExcludeFilter(AnonymousClassesFilter)
             addExcludeFilter(KotlinInternalFilter)
+            addExcludeFilter(BridgeForClassesFilter)
```
Comment (asodja) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy:54]: 💭 We filter `$$BridgeFor$$` classes for binary compatibilty check

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java (hunk 4 above)
Comment (lptr) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java:133]: I updated the PR description to include the fact that these methods are deprecated.

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java [hunk 9]
```diff
@@ -74,14 +81,16 @@
...
+    private static final Type DEFAULT_TYPE = Type.getType(DefaultValue.class);
 
     private final String projectName;
     private final Elements elements;
+    private final Types types;
 
     public PropertyUpgradeAnnotatedMethodReader(ProcessingEnvironment processingEnv) {
         this.projectName = getProjectName(processingEnv);
         this.elements = processingEnv.getElementUtils();
+        this.types = processingEnv.getTypeUtils();
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:93]: Just `ProcessingEnvironment` provided by annotation processor API is a container for all these useful types. There is no DI with annotation processor

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/filters/BridgeForClassesFilter.groovy [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import javassist.CtClass
+import org.gradle.api.Generated
+
+/**
+ * $$BridgeFor$$ classes generated for bytecode upgrade instrumentation
+ */
+class BridgeForClassesFilter implements ClassFilter {
+    @Override
+    boolean matches(CtClass ctClass) {
+        return ctClass.name.contains("\$\$BridgeFor\$\$") && ctClass.hasAnnotation(Generated.class)
```
[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/filters/BridgeForClassesFilter.groovy:29]: Do you mean to share constant between production code and build-logic? I don't think this is simply possible unless we build and use a wrapper that has this constant somewhere available. I don't think that is worth it.

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java [hunk 11]
```diff
@@ -123,9 +132,11 @@ public Collection<Result> readRequest(ExecutableElement method) {
...
                 switch (accessorSpec.accessorType) {
                     case GETTER:
-                        CallInterceptionRequest groovyPropertyRequest = createGroovyPropertyInterceptionRequest(accessorSpec, method);
+                        if (isGroovyProperty(accessorSpec.methodName)) {
+                            CallInterceptionRequest groovyPropertyRequest = createGroovyPropertyInterceptionRequest(accessorSpec, method);
+                            requests.add(groovyPropertyRequest);
+                        }
                         CallInterceptionRequest jvmGetterRequest = createJvmGetterInterceptionRequest(accessorSpec, method);
-                        requests.add(groovyPropertyRequest);
                         requests.add(jvmGetterRequest);
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:140]: Before we generated Groovy property interceptors for any method. But nnow we generate them only if they are Groovy properties, e.g. methods starts with `get*` or `is*`.

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeRequestExtra.java [hunk 12]
```diff
@@ -20,42 +20,48 @@
...
-        BinaryCompatibility binaryCompatibility
+        BinaryCompatibility binaryCompatibility,
+        ExecutableElement bridgedMethod
     ) {
         this.propertyName = propertyName;
         this.methodName = methodName;
         this.methodDescriptor = methodDescriptor;
         this.propertyType = propertyType;
-        this.isFluentSetter = isFluentSetter;
+        this.returnType = returnType;
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeRequestExtra.java:58]: Before we passed `isFluentSetter` around and we then handled this `isFluentSetter` when we generated code. But if we just translate `isFluentSetter` to a return type when reading an annotation we can get rid of this special hadling in generate code logic.
And generate code logic doesn't really care if it's a fluent setter. It just needs to know the return type.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy (hunk 8 above)
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy:54]: Makes sense, yes

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        given:
+        def givenSource = source """
+            package org.gradle.test;
+
+            import org.gradle.api.provider.Property;
+            import org.gradle.internal.instrumentation.api.annotations.ReplacesEagerProperty;
+            import org.gradle.internal.instrumentation.api.annotations.BytecodeUpgrade;
+
+            public abstract class Task {
+                @ReplacesEagerProperty(adapter = Task.TaskAdapter.class)
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy:36]: I was thinking that a good name for a class would be a method name. E.g. if it's adapter for `getMaxErrors()` a class would be `GetMaxErrorsAdapter {}`. That way it's easy to see the connection. 

We could even validate at compile time that class name follows that pattern.

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+                    }
+                }
+            }
+        """
+
+        when:
+        Compilation compilation = compile(givenSource)
+
+        then:
+        def generatedClass = source """
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy:66]: Makes sense, yes

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+                         }
+                         if (name.equals("maxErrors") && descriptor.equals("(I)Lorg/gradle/test/Task;") && (opcode == Opcodes.INVOKEVIRTUAL || opcode == Opcodes.INVOKEINTERFACE)) {
+                             mv._INVOKESTATIC(\$\$_BRIDGE_FOR\$\$_TASK\$\$_TASK_ADAPTER_TYPE, "access_set_maxErrors", "(Lorg/gradle/test/Task;I)Lorg/gradle/test/Task;");
+                             return true;
+                         }
+                     }
+                     return false;
+                 }
+            }
+        """
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy:148]: Yeah, checking the behaviour is not the easiest thing to do :)

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/filters/BridgeForClassesFilter.groovy (hunk 10 above)
[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/filters/BridgeForClassesFilter.groovy]: Added to a constant just in this class

File: platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/BytecodeUpgrade.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation that marks a method in a custom interceptor as an entry point for bytecode upgrade.
+ */
+@Retention(RetentionPolicy.CLASS)
+@Target({ElementType.METHOD})
+public @interface BytecodeUpgrade {
+}
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/annotations/BytecodeUpgrade.java:30]: Done, see https://github.com/gradle/gradle/pull/29464/commits/6d6e365cc887a070a363ddbe83ae5e1f87e66340

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (lptr) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:140]: That's good I guess. Just to make sure I understand this correctly: this fix does not really belong here, right? I'm fine with keeping it here, just want to confirm that it's unrelated to the rest of the PR.

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy (hunk 13 above)
[the remark this answers is not in the document]
Comment (lptr) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy]: An adaptor would potentially contain multiple methods, like a getter and setter, right? So it shouldn't be named after a method, but after the property it is attached to.

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy (hunk 15 above)
[the remark this answers is not in the document]
Comment (lptr) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy]: 👍

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:140]: Yes, it's just something I discovered when writing tests. So I fixed it here

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy (hunk 13 above)
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy]: I was thinking it is named after a method it's attached to: e.g.  if it's added to
```
public abstract Property<Integer> getMaxErrors()
```
it's named `GetMaxErrorsAdapter`.

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy (hunk 13 above)
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeCustomInterceptorCodeGenTest.groovy]: Fixed, now in tests we use `MaxErrorsAdapter`.

**Comment by asodja** [comment]
@bot-gradle clean

**Review by lptr — APPROVED** [review]
LGTM.

**Pull request merged**